src/main/java/net/dv8tion/jda/internal/entities/InteractionEntityBuilder.java

```diff
@@ -0,0 +1,275 @@
+    }
+
+    public Guild getOrCreateGuild(DataObject guildJson)
+    {
+        final long guildId = guildJson.getUnsignedLong("id");
+        final Guild guild = api.getGuildById(guildId);
+        if (guild != null)
+            return guild;
+
+        final Optional<DataArray> featuresArray = guildJson.optArray("features");
+        final String locale = guildJson.getString("preferred_locale", "en-US");
+
+        final DetachedGuildImpl detachedGuild = new DetachedGuildImpl(api, guildId);
+        detachedGuild.setLocale(DiscordLocale.from(locale));
+        detachedGuild.setFeatures(featuresArray.map(array ->
+                array.stream(DataArray::getString)
+                        .map(String::intern) // Prevent allocating the same feature string over and over
+                        .collect(Collectors.toSet())
+        ).orElse(Collections.emptySet()));
+
+        return detachedGuild;
+    }
+
+    public GroupChannel createGroupChannel(DataObject channelData)
+    {
+        return new DetachedGroupChannelImpl(api, channelData.getLong("id"))
+                .setLatestMessageIdLong(channelData.getLong("last_message_id", 0L))
+                .setName(channelData.getString("name", ""))
+                .setOwnerId(channelData.getLong("owner_id"))
+                .setIcon(channelData.getString("icon", null));
+    }
+
+    public GuildChannel createGuildChannel(@Nonnull Guild guild, DataObject channelData)
+    {
+        final ChannelType channelType = ChannelType.fromId(channelData.getInt("type"));
+        switch (channelType)
+        {
+        case TEXT:
+            return createTextChannel(guild, channelData);
+        case NEWS:
+            return createNewsChannel(guild, channelData);
+        case STAGE:
+            return createStageChannel(guild, channelData);
+        case VOICE:
+            return createVoiceChannel(guild, channelData);
+        case CATEGORY:
+            return createCategory(guild, channelData);
+        case FORUM:
+            return createForumChannel(guild, channelData);
+        case MEDIA:
+            return createMediaChannel(guild, channelData);
+        default:
+            LOG.debug("Cannot create channel for type " + channelData.getInt("type"));
+            return null;
+        }
+    }
+
+    public Category createCategory(@Nonnull Guild guild, DataObject json)
+    {
+        if (!guild.isDetached())
+            return guild.getCategoryById(json.getLong("id"));
+
+        final long id = json.getLong("id");
+        final DetachedCategoryImpl channel = new DetachedCategoryImpl(id, (DetachedGuildImpl) guild);
+        configureCategory(json, channel);
+        configureChannelInteractionPermissions(channel, json);
+        return channel;
+    }
+
+    public TextChannel createTextChannel(@Nonnull Guild guild, DataObject json)
+    {
+        if (!guild.isDetached())
+            return guild.getTextChannelById(json.getLong("id"));
```

Comment:
These types of getters can technically result in `null` being returned. Is that a problem? Does the consumer of this EntityBuilder know how to handle that?

Reply:
They do handle it
* `SelectMenuMentions` filters out `null`s
* `InteractionImpl` throws if the final channel is `null`
* `CommandInteractionPayloadImpl` puts it in a map as a value, the value is read by `OptionMapping` and throws if it is null